basic_pipeline/3_Source.md

```diff
@@ -0,0 +1,132 @@
+Finally, we can get to the code! 
+We will start where all the pipelines start - with the `Source` element. 
+Since this will be the first element we implement, we need to find out something more about how the Membrane Framework's elements should be implemented and some concepts associated with them.
+The first thing you need to be aware of is that `Membrane. Element` describes a specific behavior, based on the OTP GenServer's behavior.
+Our process keeps a state which is updated in callbacks. 
+We only need to provide an implementation of some callbacks in order to make our element act in the desired way. 
+The set of callbacks that can be implemented depends on the type of the elements and we will get familiar with them during the implementation of these elements.
+However, each callback is required to return a tuple of a [specific form](https://hexdocs.pm/membrane_core/Membrane.Pipeline.html#t:callback_return_t/0).
+As you can see, we are returning a status of the operation, an optional list of [actions](https://hexdocs.pm/membrane_core/Membrane.Pipeline.Action.html#t:t/0) to be performed, and 
+the updated state (which later on will be passed to the next invoked callback).
+Take your time and read about the possible actions which can be requested to be performed while returning from the callback. Their usage is crucial for the pipeline to work.
+
+As you can judge based on the structure of the project, all the elements will be put in the `lib/elements` directory. Therefore there is a place where `Source.ex` with the `Basic.Elements.Source` module's definition should be placed.
+# What makes our module a Membrane Framework's element?
+Let's start with specifying that our module will implement the `Membrane.Source` behavior as well as alias the modules which will be used later in the module's code:
+```Elixir
+# FILE: lib/elements/Source.ex
+
+defmodule Basic.Elements.Source do
+ use Membrane.Source
+ alias Membrane.Buffer
+ ...
+end
+
+```
+# Pads and options
+Later on, we will make usage of macros defined in the `Membrane.Source` module:
+```Elixir
+# FILE: lib/elements/Source.ex
+
+defmodule Basic.Elements.Source do
+ ...
+ def_options location: [type: :string, description: "Path to the file"]
+ def_output_pad :output, [caps: {Basic.Formats.Packet, type: :custom_packets}, mode: :pull]
+ ...
+end
+
+```
+
+The first macro, `def_options` allows us to define the parameters which are expected to be passed while instantiating the element. The parameters will be passed as an automatically generated structure `%Basic.Elements.Source{}`. In our case, we will have a `:location` field inside of that structure. This parameter is about to be a path to the files which will consist of input packets.
```

Comment:
Maybe some code snippet with an example of defining the child `Source`? It might be not clear to someone without experience that this defines something similar to the "constructor" of our Element.

Reply:
done